src/Features/CSharp/Portable/MakeStructFieldsWritable/CSharpMakeStructFieldsWritableDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,102 @@
+        {
+            return DiagnosticAnalyzerCategory.SemanticDocumentAnalysis;
+        }
+
+        public override bool OpenFileOnly(Workspace workspace) => true;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(compilationStartContext
+                => SymbolAnalyzer.CreateAndRegisterActions(compilationStartContext));
+        }
+
+        private sealed class SymbolAnalyzer
+        {
+            private bool _hasTypeInstanceAssigment = false;
+            private INamedTypeSymbol _namedTypeSymbol;
+            private SymbolAnalyzer(INamedTypeSymbol namedTypeSymbol)
+            {
+                _namedTypeSymbol = namedTypeSymbol;
+            }
+
+            public static void CreateAndRegisterActions(CompilationStartAnalysisContext compilationStartContext)
+            {
+                compilationStartContext.RegisterSymbolStartAction(symbolStartContext =>
+                {
+                    // We report diagnostic only if these requirements are met:
+                    // 1. The type is struct
+                    // 2. Struct contains at least one 'readonly' field
+                    // 3. Struct contains assignment to 'this' outside the scope of constructor
+                    var namedTypeSymbol = (INamedTypeSymbol)symbolStartContext.Symbol;
+
+                    // We are only interested in struct declarations
+                    if (namedTypeSymbol.TypeKind != TypeKind.Struct) return;
+                    //We check if struct contains any 'readonly' fields
+                    if (!HasReadonlyField(namedTypeSymbol)) return;
+
+                    var symbolAnalyzer = new SymbolAnalyzer(namedTypeSymbol);
+                    symbolAnalyzer.RegisterActions(symbolStartContext);
+
+                }, SymbolKind.NamedType);
+            }
+
+            private static bool HasReadonlyField(INamedTypeSymbol namedTypeSymbol)
+            {
+                return namedTypeSymbol
+                .GetMembers()
+                .OfType<IFieldSymbol>()
+                .Where(field => field.AssociatedSymbol == null)
+                .Any(field => field.IsReadOnly);
+            }
+
+            private void RegisterActions(SymbolStartAnalysisContext symbolStartContext)
+            {
+                symbolStartContext.RegisterOperationAction(AnalyzeAssignment, OperationKind.SimpleAssignment);
+                symbolStartContext.RegisterSymbolEndAction(SymbolEndAction);
+            }
+
+            private void AnalyzeAssignment(OperationAnalysisContext operationContext)
+            {
+                // We are looking for assignment to 'this' outside the constructor scope
+                var operationAssigmnent = (IAssignmentOperation)operationContext.Operation;
+                _hasTypeInstanceAssigment = operationAssigmnent.Target is IInstanceReferenceOperation instance &&
```

Comment:
>= [](start = 42, length = 1)

This should be `|=` operator, otherwise we will only honor the result from last assignment callback.

Reply:
@mavasani 
Thank you for spotting this mistake. 
After I fixed that, I found out that the analyzer is being reported for `this = new MyStruct(...)`  inside the constructor. Which is incorrect behaviour, in that case analyzer should not be reported.
Therefor I had to slightly changed the logic of analyzer. Probably you should take a look at that.

Resolved in [https://github.com/dotnet/roslyn/pull/32063/commits/ef370bb38e24f9e7e6c1e567a182ed65cfa98c1c](https://github.com/dotnet/roslyn/pull/32063/commits/ef370bb38e24f9e7e6c1e567a182ed65cfa98c1c)